### DISCUSSION TITLE
Getting intermittent net/http: TLS handshake timeout errors during sync

### DISCUSSION THREAD

**Discussion by ajrpayne** [argoproj/argo-cd#9817] [Q&A]
During the sync process, sometimes the sync process will complete, and sometimes the sync process will fail with the error 'Get "https://192.168.0.1:443/api/vi?timeout=32s": net/http: TLS handshake timeout'

We have a long database migration that occurs during the presync phase. The sync fails on objects that are part of the sync phase while the presync phase is still ongoing.

**Comment by levya4** [comment]
+1 
Experiencing the same issue.

**Comment by drackley64** [comment]
I was able to resolve this by scaling the argocd-server and argocd-repo-server deployments.

**Comment by Wojciech-Chilinski** [comment]
+1 
for me, it happens for complex applications (during peak hours it contains ~1k k8s resources)
scaling replicas as suggested by @drackley64 helped, but still happens from time to time. 
next, i want to test to increase `--kubectl-parallelism-limit` for `argocd-application-controller`

**Comment by nvpidstoob** [comment]
+1
Same problem as described by @ajrpayne with database operations in presync. Scaling argocd-server, argocd-repo-server and argocd-application-controller improved it but still occurs. Also increasing `--kubectl-parallelism-limit` for argocd-application-controller as suggested by @Wojciech-Chilinski didn't fix the problem.
Any updates on this Issue?

**Reply by ajrpayne** [reply]
@nvpidstoob I used azure aks when I had this issue. Turns out internally Azure was silently killing the connections to the control plane.

**Reply by nvpidstoob** [reply]
@ajrpayne Thanks for this Info. I also have this issues in aks. Very interessting behaviour from aks to silenty kill connections. Did you find a way solve that?

**Reply by ajrpayne** [reply]
We changed our argocd apps to not use https://kubernetes.default.svc(in-cluster), and instead use an externally routed api endpoint.

**Reply by sharninder** [reply]
@ajrpayne Do you remember how did you find out that azure was silently killing connections ? I think we're also facing the same issue but I'm unable to figure out a metric to pin point and talk to Azure about it.

**Reply by ajrpayne** [reply]
Azure support made us jump through a bunch of hoops, capturing a number of pcaps for them, before they would admit the issue is on their end.

I spammed the below script on a debug pod until I got a TLS handshake error captured by tcpdump.

while curl -ik -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" https://192.168.0.1/api; do sleep 2; done

Azure has no eta on a fix, so I continue to route outside their network.

**Reply by JRBANCEL** [reply]
If you look at the Endpoints behind the Service `https://kubernetes.default.svc`, it is the IP of the API endpoint itself.
Therefore having Argo CD use one or the other is exactly the same, so I don't see how it can fix the issue.

This might be useful: https://github.com/Azure/AKS/issues/3047#issuecomment-1721877251

**Reply by sharninder** [reply]
This is my concern too. The endpoints are the same and it shouldn't matter accessing the api server via the service or directly on the public ip. I still see the errors when I tried this solution.

**Reply by ajrpayne** [reply]
After looking at the endpoint, turns out they are indeed the same ip address. We also started using public ip prefixes for our argocd nodepools. As mentioned in https://github.com/Azure/AKS/issues/3047#issuecomment-1721877251, turns out this might be the real work-around.

**Reply by ajrpayne** [reply]
Below is what Azure support said about the issue:

Our backend team informed that they are working on a product enhancement that should significantly reduce the likelihood of this issue occurring by eliminating one of the common causes (API server load balancer timeouts).

**Reply by sharninder** [reply]
Azure has now suggested using the NAT gateway for our cluster to mitigate this issue. They say that they are aware of some bugs with their Standard load balancer and are working on fixing those, with no ETA. I'm not entirely convinced of the NAT Gateway solution but it might have a similar effect as using public IPs on our nodes.
